torchft/manager.py

```diff
@@ -943,81 +957,251 @@ def exception(self, msg: str) -> None:
         self._logger.exception(f"{self.prefix()} {msg}")
 
 
+T = TypeVar("T")
+S = TypeVar("S")
+
+
+class _SimpleFuture(torch.futures.Future[T]):
+    """
+    A simplified implementation of torch.futures.Future that wraps a value.
+
+    This class provides a minimal Future implementation that holds a pre-determined value.
+    It's primarily used as a wrapper for values in the callback chain of `_ManagedFuture`.
+    Most methods raise `RuntimeError` as they're not intended to be called.
+
+    This class is designed to be used only in specific contexts where we don't
+    want to call `value()` on the underlying `Future` as that would cause the CPU to block.
+    """
+
+    def __init__(self, value: T) -> None:
+        super().__init__()
+        self._value = value
+
+    def value(self) -> T:
+        return self._value
+
+    def then(
+        self, callback: Callable[[torch.futures.Future[T]], S]
+    ) -> torch.futures.Future[S]:
+        raise RuntimeError("should not be called")
+
+    def wait(self) -> T:
+        raise RuntimeError("should not be called")
+
+    def done(self) -> bool:
+        raise RuntimeError("should not be called")
+
+    def add_done_callback(
+        self, callback: Callable[[torch.futures.Future[T]], None]
+    ) -> None:
+        raise RuntimeError("should not be called")
+
+    def set_result(self, result: T) -> None:
+        raise RuntimeError("should not be called")
+
+    def set_exception(self, result: T) -> None:
+        raise RuntimeError("should not be called")
```

Comment:
nit: NotImplementedError instead of RuntimeError? also maybe we can make this more of a "this method is not supported" with an explanation. Could also make a custom class since this is used in so many spots